src/platform_impl/macos/window_delegate.rs

```diff
@@ -275,10 +296,12 @@ extern "C" fn window_did_move(this: &Object, _: Sel, _: id) {
 extern "C" fn window_did_change_screen(this: &Object, _: Sel, _: id) {
     trace!("Triggered `windowDidChangeScreen:`");
     with_state(this, |state| {
-        let dpi_factor = unsafe { NSWindow::backingScaleFactor(*state.ns_window) } as f64;
-        if state.previous_dpi_factor != dpi_factor {
-            state.previous_dpi_factor = dpi_factor;
-            state.emit_event(WindowEvent::HiDpiFactorChanged(dpi_factor));
+        let hidpi_factor = unsafe { NSWindow::backingScaleFactor(*state.ns_window) } as f64;
+        let ns_size = unsafe { NSWindow::frame(*state.ns_window).size };
+        let new_size = inner_size(ns_size, hidpi_factor);
+        if state.previous_dpi_factor != hidpi_factor {
+            state.previous_dpi_factor = hidpi_factor;
+            state.emit_hidpi_factor_changed_event(&mut Some(new_size));
```

Comment:
Right now, this throws away any changes the user makes to the `new_inner_rect_opt` field. We should read the value the user places in that field, and change the window's size to match it if the user wants a different size than the size suggested by the OS.

Reply:
I've set up some examples to verify that stuff is behaving correctly.

This example should set the window's physical size to `256x256` whenever the DPI factor changes, and will panic if that doesn't happen:

```rust
use winit::{
    dpi::PhysicalSize,
    event::{Event, WindowEvent},
    event_loop::{ControlFlow, EventLoop},
    window::WindowBuilder,
};

fn main() {
    let event_loop = EventLoop::new();

    let window = WindowBuilder::new()
        .with_title("A fantastic window!")
        .build(&event_loop)
        .unwrap();

    let mut assert_size = None;

    event_loop.run(move |event, _, control_flow| {
        println!("{:?}", event);

        if let Some(size) = assert_size.take() {
            assert_eq!(size, window.inner_size());
        }

        match event {
            Event::WindowEvent {
                event: WindowEvent::CloseRequested,
                window_id,
            } if window_id == window.id() => *control_flow = ControlFlow::Exit,
            Event::WindowEvent {
                event: WindowEvent::HiDpiFactorChanged{new_inner_size, ..},
                ..
            } => {
                assert_size = Some(PhysicalSize::new(256, 256));
                *new_inner_size = assert_size;
            }
            _ => *control_flow = ControlFlow::Wait,
        }
    });
}
```

This example should keep the window's physical size consistent when the DPI factor changes.

Incidentally, both of those examples panic on Windows, so it looks like I've got some bugfixing to do as well!

Reply:
Can these be turned into tests, @Osspial ?

Reply:
Ideally yes, but doing so would require a lot of work that I'm not willing to do right now. The only way I can see to automatically trigger `HiDpiFactorChanged` would be to modify the system's monitor DPI configuration, and that involves diving into the platform-specific APIs.

Reply:
@Osspial, I've tested current implementation by setting different screen resolutions, and when it should have changed size of the window, it keeps the promotions of the window.
Is it how we do want to keep it?
On the second test it performs correctly.

Reply:
@Osspial, in the first example, on hidpi change we should get a square window (256 x 256), but we get a rectangular (256 x 212). Do we want to keep it that way?

Reply:
The latter test doesn't seem correct to me. Shouldn't we be comparing the logical size, not the physical size? The physical size will always be different between two monitors with different scaling factors.

Reply:
@aleksijuvani, could you test it with the latest commit? I've tuned it all up a bit, and when I change resolution through System Preferences on my laptop both tests pass.

Reply:
Looks good to me. No extraneous events are emitted anymore. Additionally, modifying the latter test here so that `let assert_size = window.inner_size().to_logical(window.hidpi_factor())` and `assert_eq!(assert_size, window.inner_size().to_logical(window.hidpi_factor()))` will make it pass. Both of the tests pass now when dragging the window between monitors with different scaling factors.

Reply:
@aleksijuvani The expected behavior when setting `HiDpiFactorChanged` to `None` was that the window's physical size would remain consistent, which is what the second test tested for. That was the intended, and as of #1130 the exhibited, behavior.